### PR TITLE
refactor!: New `arpack_defaults()` replaces `igraph.arpack.default` and `arpack_defaults` lists

### LINKED ISSUE #714: special case the lifecycle transformation of igraph.arpack.default
since at the moment it's a list not a function

### PR DESCRIPTION
**Pull request by maelle**
…arpack_defaults() as a function

Like #762
Fix #714

### CODE REVIEW

**Comment by maelle** [comment]
@krlmlr how can I add this change (changing the default argument but also doing the "hack" inside the function to not break reverse dependencies) to functions of aaa-auto.R?

**Comment by krlmlr** [comment]
Can you please rephrase?

The checks are failing here.

**Comment by maelle** [comment]
I have made changes in some scripts manually. For the checks to pass, I'd need to update aaa-auto.R. However I do not know how to do that _not manually_. I know where aaa-auto.R content comes from but

- the "hack" with rlang is a bit longer than what's in the YAML files.
- I don't know how to update aaa-auto.R based on the YAML file.

**Comment by ntamas** [comment]
`aaa-auto.R` is generated from `tools/stimulus/aaa-auto.R.in` (which is basically empty) using Stimulus. I think it might be easier if you could simply show what you'd like to do in `aaa-auto.R` and then I can tell you what the easiest option would be.

**Comment by maelle** [comment]
For functions that have `arpack_defaults` as a default argument value, I need to change it to `arpack_defaults()`, but also to add these lines at the top of the function body

```r
  eval_try <- rlang::eval_tidy(options)
  options_value <- rlang::call_args(rlang::current_call())[["options"]]
  if (is(eval_try, "function") && as.character(options_value) == "arpack_defaults") {
    lifecycle::deprecate_soft(
      "1.5.0",
      I("arpack_defaults"),
      "arpack_defaults()",
      details = c("So the function arpack_defaults(), not an object called arpack_defaults.")
    )
    options <- arpack_defaults()
  }
```

I'll explore a bit more.

File: R/centrality.R [hunk 1]
```diff
@@ -496,8 +498,21 @@ arpack_defaults <- list(
 #' }
 #' @family arpack
 #' @export
-arpack <- function(func, extra = NULL, sym = FALSE, options = arpack_defaults,
+arpack <- function(func, extra = NULL, sym = FALSE, options = arpack_defaults(),
                    env = parent.frame(), complex = !sym) {
+
+  eval_try <- rlang::eval_tidy(options)
```
Comment (krlmlr) [inline R/centrality.R:504]: Why `eval_tidy()`? We should be able to do this with standard evaluation, forcing should be enough?

File: R/centrality.R [hunk 2]
```diff
@@ -496,8 +498,21 @@ arpack_defaults <- list(
...
+  options_value <- rlang::call_args(rlang::current_call())[["options"]]
+  if (is(eval_try, "function") && as.character(options_value) == "arpack_defaults") {
+    lifecycle::deprecate_soft(
+      "1.5.0",
```
Comment (krlmlr) [inline R/centrality.R:508]: Numbers are cheap.

**Review by krlmlr — COMMENTED** [review]
Thanks. Some functions still seem to use `igraph.arpack.default`, according to check results?

**Comment by maelle** [comment]
yes because I'm stuck, not knowing how to add the lines to the body of functions that are created automatically https://github.com/igraph/rigraph/pull/800#issuecomment-1594271629 (I won't get back to this this week)

File: R/centrality.R (hunk 1 above)
Comment (maelle) [reply R/centrality.R:504]: do you have any example I could look at?

File: R/centrality.R (hunk 2 above)
Comment (maelle) [reply R/centrality.R:508]: what do you mean?

**Comment by krlmlr** [comment]
Oh, got you. Instead of changing the body of an autogenerated function, can we wrap it instead, as done in #859?

**Comment by maelle** [comment]
In this PR I have updated types-RR.yaml, however running the makefile doesn't update aaa-auto.R, I am confused.

@krlmlr good idea, will do.

**Comment by ntamas** [comment]
> In this PR I have updated types-RR.yaml, however running the makefile doesn't update aaa-auto.R, I am confused.

Are you running `Makefile-cigraph`? This is the makefile that contains the rule that updates `aaa-auto.R`.

**Comment by maelle** [comment]
I ran `make -f Makefile-cigraph src/rinterface.c R/aaa-auto.R` as indicated in `tools/README.md`

**Review by krlmlr — CHANGES_REQUESTED** [review]
I'll tweak here too.

**Comment by krlmlr** [comment]
@maelle: Can you please take a look at the failing tests? Do you need more support?

**Comment by maelle** [comment]
My problems here are

- not knowing where to change the code for aaa-auto.R to be edited properly.
- not understanding the comments https://github.com/igraph/rigraph/pull/800#discussion_r1234748087 and https://github.com/igraph/rigraph/pull/800#discussion_r1234748138

**Review by krlmlr — APPROVED** [review]

**Comment by maelle** [comment]
thanks @krlmlr

**Pull request merged**